crates/core/src/transport/udp_transport.rs

```diff
@@ -0,0 +1,182 @@
+//! UDP transport implementation
+//!
+//! # Protocol
+//!
+//! ## Connection Establishment
+//!
+//! ### Neither peer is a gateway
+//!
+//!
+//!
+//! # Transport message types (u8)
+//!
+//! * 0: Symmetric key encrypted with our public key
+//! * 1: Acknowledgement of symmetric key - encrypted with symmetric key
+//! * 2: Message - encrypted with symmetric key
+//! * 3: Disconnect message - encrypted with symmetric key
+
+use super::*;
+use crate::transport::crypto::{TransportKeypair, TransportPublicKey};
+use crate::transport::udp_connection::UdpConnectionInfo;
+use dashmap::DashMap;
+use rand::random;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::vec::Vec;
+use tokio::net::UdpSocket;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{mpsc, RwLock};
+use tokio::task;
+
+/// The maximum size of a received UDP packet, MTU is 1500
+/// so this should be more than enough.
+const MAX_PACKET_SIZE: usize = 2048;
+
+const RECEIVE_QUEUE_SIZE: usize = 100;
+
+pub(crate) struct UdpTransport {
+    pub(super) connection_raw_packet_senders: DashMap<SocketAddr, Sender<(SocketAddr, Vec<u8>)>>,
+    pub(super) connection_info: DashMap<SocketAddr, UdpConnectionInfo>,
+    keypair: TransportKeypair,
+    listen_port: u16,
+    is_gateway: bool,
+    max_upstream_rate: BytesPerSecond,
+    send_queue: Sender<(SocketAddr, Vec<u8>)>,
+}
+
+impl UdpTransport {
+    async fn new(
+        keypair: TransportKeypair,
+        listen_port: u16,
+        is_gateway: bool,
+        max_upstream_rate: BytesPerSecond,
+    ) -> Result<Arc<RwLock<Self>>, TransportError>
+    where
+        Self: Sized,
+    {
+        // Bind the UDP socket to the specified port
+        let socket = UdpSocket::bind(("0.0.0.0", listen_port))
+            .await
+            .map_err(|e| TransportError(e.to_string()))?;
+
+        // Channel buffer is zero so senders will block until the receiver is ready,
+        // important for bandwidth limiting
+        let (send_queue, mut send_queue_receiver) = mpsc::channel(0);
```

Comment:
buffer needs to be 1, otherwise you can't send evena  single message; it will block until that one message is consumed